src/common/class-syntax.k

```diff
@@ -0,0 +1,125 @@
+require "array-syntax.k"
+require "exp-syntax.k"
+module CLASS-SYNTAX
+imports ARRAY-SYNTAX
+imports EXP-SYNTAX
+//@ \subsection{ClassDeclarations.sdf}
+syntax ClassDec ::= ClassDecHead ClassBody [klabel('ClassDec)]
+syntax ClassDec ::= EnumDec
+
+syntax ClassBody ::= "{" ClassBodyDecList "}" [klabel('ClassBody)]
+//todo optionals
+syntax ClassDecHead ::= AnnoClassModList "class" Id TypeParams Super Interfaces [prefer, klabel('ClassDecHead)]
+
+syntax ClassMod ::= Public
+                  | Private
+                  | Protected
+                  | Abstract
+                  | Final
+                  | Static
+                  | StrictFP
+syntax AnnoClassMod ::= Anno | ClassMod
+
+syntax Super ::= "extends" ClassType                    [klabel('SuperDec)]
+syntax Interfaces ::= "implements" InterfaceTypeList    [klabel('ImplementsDec)]
+
+syntax ClassBodyDec ::= InstanceInit | StaticInit | ClassMemberDec | ConstrDec
+
+syntax ClassMemberDec ::= FieldDec | ClassDec | MethodDec | InterfaceDec | SemiColon
+
+syntax SemiColon ::= ";" [klabel('SemiColon)]
+
+//@ \subsection{ConstructorDeclarations.sdf}
+//todo optionals
+syntax ConstrHead   ::=  AnnoConstrModList TypeParams Type Id "(" Params ")" Throws [prefer, klabel('ConstrDecHead)]
+                       | AnnoConstrModList TypeParams Type Id "(" Params ")"    [klabel('ConstrDecHead)]
+                       | AnnoConstrModList Type Id "(" Params ")" Throws   [klabel('ConstrDecHead)]
+                       | AnnoConstrModList Type Id "(" Params ")"    [klabel('ConstrDecHead)]
+
+syntax ConstrDec ::= ConstrHead ConstrBody                    [klabel('ConstrDec)]
+syntax ConstrBody ::= "{" ConstrInv BlockStmList "}"          [prefer, klabel('ConstrBody)]
+                    | "{" BlockStmList "}"                    [klabel('ConstrBody)]
+
+syntax ConstrInv ::= TypeArgs "this"  "(" Exps ")" ";"        [prefer, klabel('AltConstrInv)]
+syntax ConstrInv ::= "this"  "(" Exps ")" ";"                 [klabel('AltConstrInv)]
+syntax ConstrInv ::= TypeArgs "super"  "(" Exps ")" ";"       [prefer, klabel('SuperConstrInv)]
+syntax ConstrInv ::= "super"  "(" Exps ")" ";"                [klabel('SuperConstrInv)]
+syntax ConstrInv ::= Exp "." TypeArgs "super"  "(" Exps ")" ";" [prefer, klabel('QSuperConstrInv)]
+syntax ConstrInv ::= Exp "." "super"  "(" Exps ")" ";"          [klabel('QSuperConstrInv)]
+
+syntax ConstrMod ::= Public
+                  | Private
+                  | Protected
+syntax AnnoConstrMod ::= Anno | ConstrMod
+
+//@ \subsection{EnumDeclarations.sdf}
+syntax EnumDecHead ::= AnnoClassModList "enum" Id Interfaces [prefer, klabel('EnumDecHead)]
+                     | AnnoClassModList "enum" Id            [klabel('EnumDecHead)]
+
+syntax EnumDec ::= EnumDecHead EnumBody                     [klabel('EnumDec)]
+
+syntax EnumBody ::= "{" EnumConstList EnumBodyDecs"}"       [prefer, klabel('EnumBody)]
+                  | "{" EnumConstList "}"                   [klabel('EnumBody)]
```

Comment:
This is what I meant by adding macros to add the option when missing.  Don't you need a macro of the form

    rule { ECL } => { ECL emptyEnumBodyDecs }

?  You should.  If you don't, then it probably means that the JavaFront parser already does that for you, and that is not right (because our parser later on will not do that anymore).  We will have to do it with a macro.  I continue to believe that keeping the JavaFront parser unchanged is like keeping a can of worms around.